Bootstrap 5 upgrade

dstopp opened the pull request:
- Fixes #1190
- Were any Django, import pipeline, or table_schema files modified? Y/N
  - If YES:
    - Database used: opus3_test_220530
    - All Django tests pass: NA
    - FLAKE8 run on modified code: NA
- Were any JavaScript or CSS files modified? Y
  - If YES:
    - JSHINT run on all affected files: Y
    - Tested on Chrome, Firefox, Safari, Edge, and iOS: Y/NA
      (Remember to test all browser sizes)
    - Tested for race conditions (with delayed API calls if applicable): NA
- Was the documentation reviewed for necessary changes or additions? NA
  - About
  - Getting Started
  - FAQ
  - API Guide
  - Tooltips

Description of changes:

Known problems:
The close button on the modal still needs some adjusting; testing of changes is ongoing.  Testing still ongoing.

dstopp committed:
Initial changes for upgrade to bootstrap 5

update headers, add fallback files, some minor changes to allow render, not complete

dstopp committed:
Continued work on bootstrap 5 upgrade

dstopp committed:
Merge branch 'master' of https://github.com/SETI/pds-opus into 1190_bootstrap5_upgrade

dstopp committed:
Continued update for bootstrap 5; nav links, minor spacing

dstopp committed:
fix the shadow line and buttons on search tab

juzen2003 committed:
Update the fallback files for fontawesome & lodash.

juzen2003 committed:
Removed workaround code that keeps the active state for nav bar tab in
opus.js.

juzen2003 committed:
Remove empty line at line 420 in opus.js.

dstopp committed:
Merge pull request #15 from juzen2003/1190_bootstrap5_upgrade

Update the fallback files

dstopp committed:
continued updates for bootstrap 5

dstopp committed:
Merge remote-tracking branch 'origin/1190_bootstrap5_upgrade' into 1190_bootstrap5_upgrade

dstopp committed:
Continued updates to support bootstrap 5

dstopp committed:
Continued update for bootstrap 5

dstopp committed:
Continued bootstrap 5 updates

dstopp committed:
Modal and table changes

dstopp committed:
Continued changes for bootstrap 5 upgrade

dstopp committed:
On going changes for bootstrap 5

dstopp committed:
Merge branch 'master' of https://github.com/SETI/pds-opus into 1190_bootstrap5_upgrade

juzen2003 committed:
Update the behavior for help submenu to make sure submenu is properly
open and closed when interpreting preview images is clicked.

dstopp committed:
Merge pull request #17 from juzen2003/1190_bootstrap5_upgrade

Update the behavior for help submenu when clicking interpreting preview images

rfrenchseti commented on opus/application/.coveragerc:
```diff
@@ -1,12 +0,0 @@
-[run]
```
This file should not be removed

dstopp committed:
Merge branch 'master' of https://github.com/SETI/pds-opus into 1190_bootstrap5_upgrade

includes add'l fixes for close button + adding back .coveragerc file that was accidentally deleted

dstopp committed:
Merge branch 'master' of https://github.com/SETI/pds-opus into 1190_bootstrap5_upgrade

dstopp committed:
Merge remote-tracking branch 'origin/1190_bootstrap5_upgrade' into 1190_bootstrap5_upgrade

juzen2003 commented on opus/application/apps/ui/templates/ui/base.html:
```diff
@@ -6,87 +6,89 @@
+                                <a class="dropdown-item op-help-item" data-action="splash" href="#" title="View the OPUS new-user splash screen">Welcome Message</a>
+                                <a class="dropdown-item op-help-item" data-action="citing" href="#" title="How to cite OPUS in your papers or presentations">Citing OPUS</a>
+                                <a class="dropdown-item op-help-item" data-action="contact" href="#" title="Send questions or feedback to the OPUS developers">Contact Us</a>
+                           </div>
```
Add one more space to align with the opening <div>

juzen2003 reviewed (changes requested):
- The pre-programmed input dropdown is broken (ex: wavelength widget). The preprogrammed dropdown will not show up when clicking on the input.
- When opening up the download links history (at the lower right corner), the dialog window has missing (x) button and "Clear All" button. 
- When clicking on the "trash" icon in the header of any column in the browse table view, the column will be removed but the following error message will show up in the console.
```
Uncaught TypeError: Failed to execute 'getComputedStyle' on 'Window': parameter 1 is not of type 'Element'.
```

juzen2003 commented on opus/application/static_media/css/opus.css:
```diff
@@ -1233,14 +1289,23 @@ without resizing image proportionally. */
+modal is narrow */
+.op-remove-col {
+    flex: 1 0 auto !important;
+    width: 100%
```
missing ```;``` at the end.

juzen2003 commented on opus/application/static_media/css/opus.css:
```diff
@@ -1910,17 +1988,27 @@ panel are properly separated */
+    outline: none;
+    background: inherit;
+    border: 1px solid #c5c5c5;
+    color:
```
```color``` styling is not used.

juzen2003 commented on opus/application/static_media/js/browse.js:
```diff
@@ -1876,7 +1885,7 @@ var o_browse = {
 
         // check all box
-        // let addallIcon = "<button type='button' data-toggle='modal' data-target='#op-addall-to-cart-modal' " +
+        // let addallIcon = "<button type='button' data-bs-toggle='modal' data-bs-target='#op-addall-to-cart-modal' " +
```
Unused code.

dstopp committed:
Range dropdown, help

dstopp committed:
Fixed the download links history popover

dstopp committed:
Spacing and button styling for help button

Includes minor code styling fix.

dstopp committed:
Download CSV button fixes

juzen2003 committed:
- Remove the div wrapping the .dropdown-menu to work with Bootstrap.
- Remove unused css stylings.
- Update JS code on how we toggle show class to open/close dropdown.

juzen2003 committed:
Merge remote-tracking branch 'origin/1190_bootstrap5_upgrade' into 1190_bootstrap5_upgrade

juzen2003 committed:
Update the code to toggle bootstrap 5.1 dropdown via JS.

juzen2003 committed:
Fixed styling for .op-scrollable-menu (including the scrolling behaviors)

dstopp committed:
Merge pull request #18 from juzen2003/1190_bootstrap5_upgrade

Update the code to toggle bootstrap 5.1 dropdown via JS.

dstopp committed:
Merge

dstopp committed:
Upgrade to bootstrap 5.20 

attempt to fix dropdown bugs in 5.1.x

dstopp committed:
Position and class fix

juzen2003 committed:
Removed the opening help menu when clicking an item in edit metadata menu
from "+" in browse table view.

juzen2003 committed:
Removed commented out code in base.html

dstopp committed:
Merge pull request #19 from juzen2003/1190_bootstrap5_upgrade

1190 bootstrap5 upgrade - Removed help menu when clicking on add column menu in browse table view

dstopp committed:
fix for the dropdown on the edit-field-tool

dstopp committed:
Fix open of help menu on mini-menu of gallery

add data-bs-toggle="dropdown"

rfrenchseti reviewed (changes requested):
A few of these are cosmetic and may not be worth fixing (or we like the new versions better) but many of the cosmetic changes I think look worse than the old version. Others are just plain bugs.

- On search widgets, the vertical bar to the left of the minimize/close buttons is now very pale.
- On the other hand, on search widgets that have an "-OR-", the "-" beside the OR are now dark black instead of the same grey as the word "OR".
- On search widgets, the title bar is not a pure grey instead of blue-grey
- On a preprogrammed ranges dropdown, the headers are now in grey text, which makes them harder to read. The text under each header is also grey.
- The headings under the Help menu are grey
- When clicking on Interpreting Preview Images under Help, the old behavior was animated and the new behavior just pops into existence
- When the browser is shrunk, the Reset to Default button in Select Metadata becomes misplaced
- When you add a sort field by pressing "+" next to "Sort by", the main help menu pops up in a weird place and never goes away
- As you shrink the browser vertically, you can make the slideshow dialog move off the top of the screen
- The hamburger menu entries on the Browse tab are grey
- When in table view, if you start a range with Shift+click or "Start range" from a hamburger menu, the green animated square is in the wrong place and the checkbox and hamburger move down some
- When in table view with enough columns that you need a horizontal scrollbar, if you scroll to the right and then show the "+ Trash +" dropdown, it's in the wrong place
- There is a very slight gap above the table header in table view such that if you scroll up the data you can see parts of white characters in the gap
- If you activate the hamburger menu, then click the menu's titlebar to close it, the main Help menu shows up and won't go away
- If you pull up the Detail tab for one of the observations with multiple preview images (e.g. UVIS/VIMS/RSS occultation profiles), no previews are shown at all. This might be a merge problem.
- I see the following error in the console whenever I switch to the Detail tab:

```
jquery-3.6.0.min.js:2 jQuery.Deferred exception: $(...).flexslider is not a function TypeError: $(...).flexslider is not a function
    at Array.<anonymous> (http://127.0.0.1:8000/static_media/js/detail.js?version=5c7f916830a2c3fac2966bca03e1ab8f4d84a0f5:135:42)
    at e (https://code.jquery.com/jquery-3.6.0.min.js:2:30038)
    at t (https://code.jquery.com/jquery-3.6.0.min.js:2:30340) undefined
```

- There is not as much vertical margin between the horizontal blue divider at the top of the screen and the View Table line (browse or cart) as there used to be. I think it looks asymmetric and cramped now.

dstopp committed:
Merge branch 'master' of https://github.com/SETI/pds-opus into 1190_bootstrap5_upgrade

dstopp committed:
minor changes to address comments

dstopp committed:
fix multiple preview images

missing include file from bad merge

dstopp committed:
Minor updates for detail tab and headings

dstopp committed:
Fix for the hamburger menus

stop help menu; change color

dstopp committed:
Most of the remaining minor changes

dstopp committed:
Merge branch 'master' of https://github.com/SETI/pds-opus into 1190_bootstrap5_upgrade

# Conflicts:
#	opus/application/apps/cart/templates/cart/cart.html

dstopp committed:
Merge branch 'master' of https://github.com/SETI/pds-opus into 1190_bootstrap5_upgrade

# Conflicts:
#	opus/application/test_api/test_results_api.py

dstopp committed:
Merge branch 'master' of https://github.com/SETI/pds-opus into 1190_bootstrap5_upgrade

dstopp committed:
remove temp files

dstopp committed:
Fix some merge issues

divider line on search tab; cart.html issues

juzen2003 committed:
Remove 1px gap on top of the table header when scolling the browse
table view. (Caused by the 1px top border of the tr)

dstopp committed:
Merge pull request #20 from juzen2003/1190_bootstrap5_upgrade

Remove 1px gap on top of the table header when scrolling the browse

dstopp committed:
fix merge issue

rfrenchseti commented on opus/application/static_media/css/opus.css:
```diff
@@ -40,8 +43,54 @@ body.modal-open .navbar {
+
 /* Make the main nav bar just a little taller than normal (style says 8px) */
-#op-main-nav {
+#op-main-nav  {
```
Extra space?

dstopp replied:
I removed the extra space; re: infinite_scroll version 3.0.5, that is the current version we are testing with.  4.0.1 is not yet implemented and should be done after the bs5 work is complete.
re: jquery, fontawesome, popper and lodash - removed unused/outdated files.

rfrenchseti commented on opus/application/static_media/js/browse.js:
```diff
@@ -2729,15 +2737,15 @@ var o_browse = {
         let toolElem = `${tab} .op-edit-field-tool`;
-        let left = $(targetElem).position().left +
-                    $(`${tab} .op-data-table-view .table`).offset().left;
+        let left = $(targetElem).position().left ;/*+
```
Why is this commented out?

dstopp replied:
removed commented out code

rfrenchseti commented on opus/application/static_media/js/search.js:
```diff
@@ -276,13 +266,15 @@ var o_search = {
 
                 if (preprogrammedRangesDropdown.hasClass("show")) {
-                    inputToTriggerDropdown.dropdown("toggle");
+                    // preprogrammedRangesDropdown.toggleClass("show");
```
Remove comments?

dstopp replied:
removed commented out code

rfrenchseti reviewed (changes requested):
Shouldn't you remove infinite_scroll version 3.0.5? Same for old versions of jquery.

dstopp committed:
Address pull comments

removed outdated/unused files; address minor code style issues.

juzen2003 commented on opus/application/static_media/js/browse.js:
```diff
@@ -2729,15 +2737,14 @@ var o_browse = {
+            if ($(".op-cart-details:visible").length > 0) {
                 // need to calculate a new offset if the table has scrolled left behind the details window
-                left -= $(".cart_details").outerWidth() + $(".cart_details").offset().left;
+                //left -= $(".op-cart-details").outerWidth() + $(".op-cart-details").offset().left;
```
This if statement block can be removed since the code is commented out.

rfrenchseti replied:
It doesn't look like this comment was ever acted on

juzen2003 reviewed (changes requested):
When clicking the items under the following two dropdown menus, there will be error messages in the console from dropdown.js (from bootstrap)
- "+" to add a new field in "sort by" row
- "+ Add new field here" to add a new field in browse view metadata modal 
The functionality is good, just the error messages will always show up when an item is selected in either of the menus.

juzen2003 committed:
Remove data-bs-toggle="dropdown" from "#op-add-sort-metadata" & its
".dropdown-item" to remove the error message from dropdown.js when
clicking an dropdown item.

juzen2003 committed:
Remove data-bs-toggle="dropdown" from "op-metadata-details-tools"

dstopp committed:
Merge pull request #21 from juzen2003/1190_bootstrap5_upgrade

Fixed the console error messages from dropdown.js

dstopp committed:
Merge branch 'master' of https://github.com/SETI/pds-opus into 1190_bootstrap5_upgrade

dstopp committed:
Merge branch 'master' of https://github.com/SETI/pds-opus into 1190_bootstrap5_upgrade

# Conflicts:
#	opus/application/static_media/css/opus.css
#	opus/application/static_media/js/browse.js

juzen2003 committed:
Remove data-bs-toggle="dropdown" from .op-metadatabox-tooltip to avoid
console error message when clicking hamburger in metadata modal.

juzen2003 committed:
Remove data-bs-toggle="dropdown" on #op-obs-menu and all its dropdown-items
to avoid console error messages when clicking any of the dropdown menu item.

dstopp committed:
Merge pull request #22 from juzen2003/1190_bootstrap5_upgrade

Remove console error messages when clicking hamburger menu or items in the menu

dstopp committed:
Minor merge issue + attempt to fix download window

download link window is not yet fixed with these changes.

juzen2003 committed:
Fixed the behaviors of the download links popover.

dstopp committed:
Merge pull request #23 from juzen2003/1190_bootstrap5_upgrade

Fixed the behaviors of the download links popover.

dstopp committed:
remove console.log messages

dstopp committed:
Merge branch 'master' of https://github.com/SETI/pds-opus into 1190_bootstrap5_upgrade

dstopp committed:
remove comment

dstopp committed:
Merge remote-tracking branch 'origin/master' into 1190_bootstrap5_upgrade

rfrenchseti commented on opus/application/test_api/test_results_api.py:
```diff
@@ -385,7 +385,7 @@ def test__api_metadata_vg_iss_2_s_c4360845_default_html(self):
         "[test_results_api.py] /api/metadata_v2: vg-iss-2-s-c4360845 default html private"
         url = '/__api/metadata/vg-iss-2-s-c4360845.html'
-        expected = b'<ul id="detail__data_general_constraints" class="op-detail-list">\n<li class="op-detail-entry">\n<i class="fas fa-info-circle op-detail-entry-icon op-detail-metadata-tooltip" data-toggle="tooltip"\ntitle="'
+        expected = b'<ul id="detail__data_general_constraints" class="op-detail-list">\n<li class="op-detail-entry">\n<i class="fas fa-info-circle op-detail-entry-icon op-detail-metadata-tooltip" data-bs-toggle="tooltip"\ntitle="'
```
Bad merge

dstopp committed:
Update tests for bs5

dstopp committed:
Update tests to remove whitespace

dstopp committed:
Merge branch 'master' of https://github.com/SETI/pds-opus into 1190_bootstrap5_upgrade

dstopp committed:
Windows only requirements

rfrenchseti reviewed (changes requested):
- In the Help menu, the submenus for Interpreting Preview Images don't link to anything when you click on them.
- Bootstrap 5.2.2 is the current version but we are using 5.2.0.
- opus.css has inconsistent whitespace - some indents are 2 chars and some are 4.
- search.js and widgets.js use the symbol `bootstrap` but it wasn't declared as a global at the top of the file, thus causing errors in JSHINT
- ui/templates/ui/header.html still has references to bootstrap 5.1.3. This part of the file is used when creating PDF versions of help files, and this is why PDF creation is currently failing.
- Pull up an observation in the slideshow. Use the edit mode to add a metadata field. The UI will freeze completely. Or sometimes it will refresh the display but not actually let you add a field. This only happens from the browse tab, not the cart tab.
- When you "start add range to cart" from the table view, there is no green box indicator on the left

dstopp committed:
Update bootstrap to 5.2.2 

fix help menu, whitespace in opus.css, and hint issues with search.js and widgets.js

dstopp committed:
Fix for jshint

dstopp committed:
Fix for green box on multi select

dstopp committed:
Freeze issue w/slideshow

this is a timing issue; i have shortened the time as it was taking too long to start and then leaving it in a hung state.

dstopp committed:
Set spinner delay back to original; hide download history

dstopp committed:
Freeze, error messages, spinner

Fixes the freeze on edit of metadata details in the slide view,  removes error messages in console and fixes the spinner for the edit of metadata in slide view so it shows.

rfrenchseti reviewed (changes requested):
- The current master branch is not merged in.
- 6 tests are failing.
- Can you just remove the infinite-scroll-4.0.1.pkg since we aren't using it yet?
- I was still able to get the slideshow add/remove metadata to freeze up, but it's harder to do now. We can defer this to a future pull.
- HOWEVER, with the current version, delete a column while in table view always freezes now. This didn't happen before.

juzen2003 reviewed (changes requested):
- Add multiple observations to cart, click view gallery & view table, this console error message will show up whenever we switch between these two views in cart:
```
Uncaught TypeError: Cannot read properties of undefined (reading 'backdrop')
```
- In search tab, mult values are duplicated for the Ground-based category in instrument name widget, this can be fixed by merging current master branch.

dstopp committed:
Merge branch 'master' of https://github.com/SETI/pds-opus into 1190_bootstrap5_upgrade

dstopp committed:
Update metadata selector test responses

dstopp committed:
fix error message in cart on change of view tabs

dstopp committed:
Fix column delete freeze, remove unused file

dstopp committed:
Merge branch 'master' of https://github.com/SETI/pds-opus into 1190_bootstrap5_upgrade

juzen2003 reviewed (approved):
- Database used: re-import
- All Django tests pass: Y
- JSHINT run on all affected files: NA
- FLAKE8 run on all affected files: Y
- Code coverage run: Y
- Tested on Chrome, Firefox, Safari, and iOS: NA
  (Remember to test all browser sizes)
- Tested for race conditions (with delayed API calls if applicable): NA
